rids/rids.py

```diff
@@ -15,113 +15,111 @@
 # limitations under the License.
 
 
+
 """
 Main entry point for RIDS, a Remote Intrusion Detection System.
 """
 
-
 import ipaddress
 import json
-import os
-import subprocess
-import sys
-import urllib.request
+import logging
+import threading
+import urllib
+import queue
 
 from absl import app
 from absl import flags
 
+from rids import config
 from rids import network_capture
+from rids.iocs import iocs
+from rids import rules
 
 FLAGS = flags.FLAGS
 flags.DEFINE_string('host_ip', None,
                     'The IP of the host process, for ignoring direct requests')
-flags.DEFINE_string('config_path', 'config.json',
-                    'Path where IOC configuration can be found; uses config.json by default')
-
-
-def retrieve_bad_ips():
-  bad_ips = {}
-  with open(FLAGS.config_path, 'r') as f:
-    config = json.load(f)
-    for source in config['ioc_sources']:
-      # Download a fresh version of each IOC source in the config.
-      with urllib.request.urlopen(source['url']) as ioc_data:
-        # TODO check format of ioc source, branch to different parsers
-        # For now, the only source format is newline-seprated bad IPv4 addresses
-        # TODO perform parsing in a separate function from config handling
-        for line in ioc_data.readlines():
-          line = line.strip().decode('utf-8')
-          if not line: continue
-          bad_ip = str(ipaddress.ip_address(line))
-          if bad_ip not in bad_ips:
-            bad_ips[bad_ip] = [source['name']]
-          else:
-            bad_ips[bad_ip].append(source['name'])
-  return bad_ips
-
-
-def get_external_ip():
-  with urllib.request.urlopen('https://ipinfo.io/ip') as my_ip:
-    return my_ip.read().decode("utf-8").strip()
+flags.DEFINE_string('config_path', '/etc/rids/config.json',
+                    'file path indicating where IOC configuration can be found')
+flags.DEFINE_string('eventlog_path', '/var/rids/events.log',
```

Comment:
This is not a safe default because it may cause undesirable changes to the user's system. Probably better to leave it empty. The config is fine because it's read-only.

(Sorry, I've been bitten by those flags too many times)

Reply:
We need to write the events data somewhere, /var is the standard location in Linux for these kinds of log files, and by definition /var is where files that are expected to change go.  The rids/ subdirectory makes it unlikely to collide with other systems.

We could check that the /var path exists and is writeable at install time, maybe?  Do you have a better suggestion for where to write this instead of /var?  I don't think /tmp is better, and the user's home path may not exist (e.g. on our DO installs the default user is root and there is no /home/root).  The repo path isn't a good candidate, either.  I could make it None here and only pass it from the service config but we'd still have to decide where to put it.

Or, are you suggesting that just the python definition should be None and using this path is fine if configured elsewhere?  I can put it in the service config where RIDS execution is specified.

Reply:
I'm sorry I wasn't clear. The path is fine, but let's specify put that in the install script, rather than hardcode it here. Perhaps the config_path should default to local directory, for more convenient development.

Reply:
I put it in the service/systemctl config, the install script doesn't run RIDS.  I also added a "use cwd if unset" condition in the python code.